### ISSUE TITLE
RFC 9266: Channel Bindings for TLS 1.3 support

### ISSUE THREAD

**Issue by Neustradamus** [dotnet/runtime#73118]
### Description

Can you add the support of RFC 9266: Channel Bindings for TLS 1.3?
- https://datatracker.ietf.org/doc/html/rfc9266 

Little details, to know easily:
- tls-unique for TLS =< 1.2
- tls-exporter for TLS = 1.3 

Thanks in advance.

Linked to:
- https://github.com/dotnet/runtime/search?q=tls-unique

### Reproduction Steps

Please read the RFC9266.

### Expected behavior

Please read the RFC9266.

### Actual behavior

Please read the RFC9266.

**Comment by Neustradamus** [comment]
@wfurt, @aik-jahoda: What do you think?

**Comment by Neustradamus** [comment]
@wfurt, @aik-jahoda, @filipnavara, @hughbe: What do you think?

I have added some info in the main ticket.

**Comment by wfurt** [comment]
not in 7.0. We may look into it in 8.0. On Windows this comes from schannel. We will need to check what is supported there.

**Comment by Neustradamus** [comment]
Dear @dotnet team,

Have you progressed on it?

**Comment by wfurt** [comment]
no, but we may during 9.0. On Windows, this could just work as we depends on schannel implementation. It will need some more research and testing. What is your intended use case @Neustradamus ?

**Comment by Neustradamus** [comment]
Dear all @dotnet team,

Any news of this security missing problem?

Thanks in advance.

Where I can send an email to security Microsoft team?

**Comment by jstedfast** [comment]
@wfurt

Hi, I'm the author of MailKit and @Neustradamus has asked me to comment here, although I don't think my comment will go the way he expects.

It appears that @Neustradamus [goes around to every single project on GitHub that supports any mode of SASL authentication](https://github.com/scram-sasl/info/issues/1) and requests that they implement support for all of the (numerous) SCRAM-SHA-* authentication mechanisms combined with the channel binding support that the `-PLUS` variants require.

When he asked me to implement `tls-export` channel binding support for the SCRAM mechanisms, I told him that I couldn't because .NET's SslStream didn't support it which is how this issue got opened.

Foolishly, I wasted a ton of time implementing all of the SCRAM mechanisms and channel binding support for `tls-unique` and `tls-server-end-point` several years ago under the assumption that he had filed these feature requests for MailKit because he needed them (as opposed to him using MailKit and dozens of other hapless projects as evidence that there is wide support for it).

Now I see that not a single human being is asking for SCRAM on *any* of these products other than @Neustradamus himself, which leads me to conclude that there is absolutely no organic demand for it anywhere.

I'm sure @Neustradamus has all of the best intentions, but I wish that he realized how much wasted effort he is causing for developers across a wide variety of projects that might be better off focusing their effort on other things that matter more.

Many (if not all) of these products that have implemented SCRAM support with channel bindings have no unit tests because there are no real-world products that anyone can actually test against, or if they *do* have unit tests, they can only do what MailKit has done and test against the 1 scenario that is provided in the specification.

In conclusion, this is all about checking a checkbox and nothing more as far as I can tell.

My recommendation would be to wait and see if there is any actual organic demand for it. That's my plan from here on out.

**Comment by Neustradamus** [comment]
@jstedfast: SCRAM-SHA-1, SCRAM-SHA-1-PLUS, SCRAM-SHA-256, SCRAM-SHA-256-PLUS have been added in GNU SASL by @jas4711, the author of your latest link.

You can see a lot of information about SCRAM and Channel Binding and a list of supported projects:
- https://github.com/scram-sasl/info/issues/1

Old and unsecure mechanisms like CRAM-MD5 and DIGEST-MD5 have been removed from Cyrus SASL and Cyrus IMAP in master code.

You can see that Cyrus SASL/IMAP, Dovecot, Exim support SCRAM-SHA-*-PLUS and several other softwares, look the uncomplete list.

**Comment by wfurt** [comment]
thanks for the feedback @jstedfast. I'll ping Windows TLS team to see what is their take on it.

**Comment by Neustradamus** [comment]
@wfurt: Thanks for your ping to Windows TLS team!

I hope a security solution soon.

It is linked to Microsoft Channel Binding:
- https://www.google.com/search?q=microsoft+channel+binding

It is needed to have a compatibility with:
- tls-unique for TLS =< 1.2 (RFC5929)
- tls-server-end-point (RFC5929)
- tls-exporter for TLS = 1.3 (RFC9266)

**Comment by Neustradamus** [comment]
Any progress on it?

**Comment by wfurt** [comment]
no, It is certainly out of scope for 10.0.

**Comment by Neustradamus** [comment]
@wfurt: Have you progressed on it?

@paraddise has spoken about tls-exporter previously.

**Comment by Neustradamus** [comment]
Dear @dotnet team,

Have you progressed on it?

For example:

@ammarheidari (@araditc team) always waits you to add "tls-exporter" support in Artalk.Xmpp library:
- https://github.com/araditc/Artalk.Xmpp/issues/1
- https://github.com/araditc/Artalk.Xmpp/issues/29
- https://github.com/araditc/Artalk.Xmpp/issues/30
- https://github.com/araditc/Artalk.Xmpp/issues/31
- https://github.com/araditc/Artalk.Xmpp/issues/32
- https://github.com/araditc/Artalk.Xmpp/issues/33
- https://github.com/araditc/Artalk.Xmpp/issues/37

@coder2000 (@ubiety team) always waits you to add "tls-exporter" support in Ubiety.Scram.Core library:
- https://github.com/ubiety/Ubiety.Scram.Core/issues/6
- https://github.com/ubiety/Ubiety.Scram.Core/issues/5

Linked to:
- https://github.com/dotnet/runtime/issues/112529
